README.md
用了 csv 的資料去把資料放進熱區圖中顯示。

1. https://github.com/CSSEGISandData/COVID-19

2. [PTT 網友與 g0v 協作的本土足跡](https://g0v.hackmd.io/@yitzu/covid-19/https%3A%2F%2Fpse.is%2F3gprqa)

3. [CSV data source (含經緯度以及 geojson 區塊)](https://docs.google.com/spreadsheets/d/e/2PACX-1vTLkgDcQOvZO_UVnvWzF06vkX8YfD_TD3vdjTCBsnYJD8nCumrvIQxaQAAfdsjxEI1J12VD8m-NDgJW/pub?gid=582904711&single=true&output=csv)

4. [CDC 開放資料](https://od.cdc.gov.tw/eic/Age_County_Gender_19Cov.json)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.9.0](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.8.0...v1.9.0) (2021-06-11)


### Features

* add twcountry ([1e7e132](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/1e7e1321a765ae4daaaca25d48339a46a5bddfba))


### Bug Fixes

* **chart.js:** sma ([8ae093c](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/8ae093c6ddc68e7ce3925739fa7ecac3413bcf81))

## [1.8.0](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.7.1...v1.8.0) (2021-05-29)


### Features

* add sma ([6ad3bbf](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/6ad3bbf140c937ddcd22453249209d176728bb8f))
* update docs' files ([d4a5c62](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/d4a5c62fec16587e8d0f114ab7077fb0bcc9ea84))
* Update markerClusterGroup and Chart ([fb5d9bb](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/fb5d9bbddf61bac8c181adacdbb819fb9792d38c))


### Bug Fixes

* taiwan ([24507c4](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/24507c48746361f706b6280806eb7cc9354aea10))

### [1.7.1](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.7.0...v1.7.1) (2020-06-09)


### Bug Fixes

* cdc data ([02e92ad](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/02e92add248391d88f45fc887b4468d067549c5f))
* 英國資料抓錯 ([9f7b304](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/9f7b304f928b1cba729c991204d7f9ff7b565755))
* 負值歸零 ([32295b4](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/32295b40da3e85b58d050f1130a76e7d1d45f506))

## [1.7.0](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.6.0...v1.7.0) (2020-05-05)


### Features

* removeFbclid ([8439eb1](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/8439eb156db1a8cdea66d6d921c4b9b573de9e2d))
* 更新美國資料，顯示大於 500 確診數的區域 ([445655e](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/445655e81e5069363cdeac9e7b5eabb5ea7c12da))

## [1.6.0](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.5.1...v1.6.0) (2020-04-26)


### Features

* **dataTable.js:** 預設隱藏 DataTable ([b5484f7](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/b5484f7e5edc1763caf55817387c0c51ac5fc594))

### [1.5.1](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.5.0...v1.5.1) (2020-04-26)


### Bug Fixes

* chartCountry undefined ([c2f1060](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/c2f1060e176c2d173052c849882a6f008984f50f))
* select 樣式 ([12aebad](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/12aebad5ec904ba9805736c2ce0ffc45f59d89b5))

## [1.5.0](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.4.0...v1.5.0) (2020-04-26)


### Features

* 新增增加量及百分比 ([fc8545c](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/fc8545ca2dc4538b56372f5627c7a9c667df0ad5))


### Bug Fixes

* 亂數點位計算方法 ([b25a699](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/b25a6999df3057d59c93dc92d388f22ab7c4b352))
* 修正參數 ([fd372c1](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/fd372c101218f2532514bd4184fa9ca84e0a7159))
* 樣式 ([81c5957](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/81c5957fe23d45e0ba631267bdd5c665dfc32931))

## [1.4.0](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.3.0...v1.4.0) (2020-04-14)


### Features

* **DataTable:** 新增 DataTable ([57d3bcd](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/57d3bcd46961c7f68ffbd796409953d3750d62d4))


### Bug Fixes

* change data source ([cb5886d](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/cb5886d89f3e0edb0f03bc0eaec25885dec76d43))

## [1.3.0](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.2.0...v1.3.0) (2020-03-28)


### Features

* auto generate new data in github action ([7604594](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/76045948d3d88249b8c2791ea6bddd1bf58a8790))
* 每日增加量 ([ed8bde6](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/ed8bde6816de9203cbcb04fd8930a1de90aa0d24))


### Bug Fixes

* add timeout error ([04b00e8](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/04b00e867735657d699b47b2be704e3d4e8d8d21))
* auto resize chart ([b91cb92](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/b91cb920645587fcb95a42baf7162ce616f2447f))
* disable animation ([798f0ab](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/798f0ab777d8099ec889753349e587636c7d9c67))
* show taiwan only ([cefc6c2](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/cefc6c247445d1e3218aa811c1bf8a01fbe0862f))
* ux of mobile ([525892d](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/525892d000b397f4d3fd461ad6d728ff36ba75e3))
* 顏色統一 ([13147ee](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/13147ee79a3f8798f937a97b56a532e56b142165))

## [1.2.0](https://github.com/LinYenCheng/virus-and-where-to-find-them/compare/v1.1.2...v1.2.0) (2020-03-26)


### Features

* add cache ([0a42bbd](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/0a42bbd167dc928d18bc66f47221ea15e798ba15))
* add loading spinner and refactor ([0142660](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/0142660320f02c25cd0577f52adb8302c234299f))
* use cdc's json ([6551924](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/655192499988c884a86744efd59a19439a23e97b))
* 加入各國資料 ([27d5df2](https://github.com/LinYenCheng/virus-and-where-to-find-them/commit/27d5df22c80a3cc93b6808f30be9542bf3959abb))

generateJSON.js
"use strict";

const fs = require("fs");
const axios = require("axios");
const dayjs = require("dayjs");
const csvToJSON = require("csvjson-csv2json");

function writeResToJSON(res, fileName) {
  let data = JSON.stringify(res.data);
  fs.writeFileSync(`./data/${fileName}.json`, data);
}

function writeResToCSV(res, fileName) {
  fs.writeFileSync(`./data/${fileName}.csv`, res.data);
  const data = fs.readFileSync(`./data/${fileName}.csv`, { encoding: "utf8" });
  const json = csvToJSON(data).filter((elm) => {
    const { end } = elm;
    const date1 = dayjs(end);
    const date2 = dayjs();
    const hours = date2.diff(date1, "hours");
    const days = Math.floor(hours / 24);
    return days < 14;
  });
  fs.writeFileSync(
    `./data/${fileName}.json`,
    JSON.stringify(
      json.map((elm) => {
        const tempElement = { ...elm };
        delete tempElement.GeoJSON;
        return tempElement;
      })
    )
  );
}

function writePartialTimeSeriesForAPI(finalTimeSeriesData) {
  var async = require("async");
  async.each(
    finalTimeSeriesData,
    function (file, callback) {
      fs.writeFile(
        "./docs/data/" + file.region.toLowerCase().replace("*", "") + ".json",
        JSON.stringify([{ ...file }], null, 4),
        function (err) {
          if (err) {
            console.log(err);
          } else {
            console.log(file.region + ".json was updated.");
          }
          callback();
        }
      );
    },
    function (err) {
      if (err) {
        // One of the iterations produced an error.
        // All processing will now stop.
        console.log("A file failed to process");
      } else {
        console.log("All files have been processed successfully");
      }
    }
  );
}

//var csv is the CSV file with headers
function csv2JSON(csv) {
  const lines = csv.split("\n");
  const result = [];
  const headers = lines[0].split(",");

  for (let i = 1; i < lines.length; i++) {
    if (!lines[i]) continue;
    const obj = {};
    const currentLine = lines[i]
      .replace("Korea, South", "S.Korea")
      .replace("Bonaire, Sint Eustatius and Saba", "Sint Eustatius and Saba")
      .replace(/\"/g, "")
      .split(",");

    for (let j = 0; j < headers.length; j++) {
      obj[headers[j]] = currentLine[j];
    }
    result.push(obj);
  }
  return result;
}

axios
  // .get("https://od.cdc.gov.tw/eic/Weekly_Age_County_Gender_19CoV.json")
  .get("https://od.cdc.gov.tw/eic/Age_County_Gender_19Cov.json")
  .then((res) => {
    writeResToJSON(res, "taiwan");
  });

axios
  .get(
    "https://docs.google.com/spreadsheets/d/e/2PACX-1vTLkgDcQOvZO_UVnvWzF06vkX8YfD_TD3vdjTCBsnYJD8nCumrvIQxaQAAfdsjxEI1J12VD8m-NDgJW/pub?gid=582904711&single=true&output=csv"
  )
  .then((res) => {
    writeResToCSV(res, "covid-activity");
  });

const instanceGithub = axios.create({
  baseURL:
    "https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_time_series",
  timeout: 5000,
  headers: { "X-Custom-Header": "foobar" },
});

function removeProperty(obj) {
  const tempElm = JSON.parse(JSON.stringify(obj));
  delete tempElm["Country/Region"];
  delete tempElm["Province/State"];
  delete tempElm["Lat"];
  delete tempElm["Long"];
  return tempElm;
}

function getLastData(obj) {
  //   const nowDate = new Date();
  //   return (
  //     obj[`${nowDate.getMonth() + 1}/${nowDate.getDate() - 1}/20`] ||
  //     obj[`${nowDate.getMonth() + 1}/${nowDate.getDate() - 2}/20`]
  //   );
  const keys = Object.keys(obj);
  const last = keys[keys.length - 1];
  return obj[last];
}

function getLastDiffData(obj) {
  const nowDate = new Date();
  let intLastDiff = 0;
  const keys = Object.keys(obj);
  const last = keys[keys.length - 1];
  const last1 = keys[keys.length - 2];
  if (last1) {
    intLastDiff = obj[last] - obj[last1];
  }

  //   if (obj[`${nowDate.getMonth() + 1}/${nowDate.getDate() - 1}/20`]) {
  //     intLastDiff =
  //       parseInt(obj[`${nowDate.getMonth() + 1}/${nowDate.getDate() - 1}/20`]) -
  //       parseInt(obj[`${nowDate.getMonth() + 1}/${nowDate.getDate() - 2}/20`]);
  //   } else {
  //     intLastDiff =
  //       parseInt(obj[`${nowDate.getMonth() + 1}/${nowDate.getDate() - 2}/20`]) -
  //       parseInt(obj[`${nowDate.getMonth() + 1}/${nowDate.getDate() - 3}/20`]);
  //   }
  return intLastDiff;
}

function sumObjectsByKey(...objs) {
  return objs.reduce((a, b) => {
    for (let k in b) {
      if (b.hasOwnProperty(k)) a[k] = (parseInt(a[k]) || 0) + parseInt(b[k]);
    }
    return a;
  }, {});
}

axios
  .all([
    instanceGithub.get("/time_series_covid19_confirmed_global.csv"),
    instanceGithub.get("/time_series_covid19_recovered_global.csv"),
    instanceGithub.get("/time_series_covid19_deaths_global.csv"),
  ])
  .then(
    axios.spread((resCases, resRecovered, resDeaths) => {
      const jsonCases = csv2JSON(resCases.data);
      const jsonRecovered = csv2JSON(resRecovered.data);
      const jsonDeaths = csv2JSON(resDeaths.data);

      let sumCases;
      let sumDeaths;
      let sumRecovered;

      const finalTimeSeriesData = jsonCases
        .map((elm) => {
          let cases;
          let deaths;
          let recovered;
          let strRegion = "";

          if (elm["Province/State"] === "") {
            cases = elm;
            deaths = jsonDeaths.find(
              (elmIn) =>
                elmIn["Country/Region"] === elm["Country/Region"] &&
                elmIn["Province/State"] === ""
            );
            recovered = jsonRecovered.find(
              (elmIn) =>
                elmIn["Country/Region"] === elm["Country/Region"] &&
                elmIn["Province/State"] === ""
            );
            strRegion = elm["Country/Region"];
          } else {
            cases = jsonCases.find(
              (elmIn) =>
                elmIn["Province/State"] === elm["Province/State"] &&
                elmIn["Province/State"] !== ""
            );
            deaths = jsonDeaths.find(
              (elmIn) =>
                elmIn["Province/State"] === elm["Province/State"] &&
                elmIn["Province/State"] !== ""
            );
            recovered = jsonRecovered.find(
              (elmIn) =>
                elmIn["Province/State"] === elm["Province/State"] &&
                elmIn["Province/State"] !== ""
            );
            if (deaths === undefined || recovered === undefined) {
              console.log(elm["Province/State"]);
              return false;
            }

            strRegion = `${elm["Province/State"]}-${elm["Country/Region"]}`;
          }

          cases = removeProperty(elm);
          deaths = removeProperty(deaths);
          recovered = removeProperty(recovered);

          if (!sumCases) {
            sumCases = cases;
            sumDeaths = deaths;
            sumRecovered = recovered;
          } else {
            sumCases = sumObjectsByKey(sumCases, cases);
            sumDeaths = sumObjectsByKey(sumDeaths, deaths);
            sumRecovered = sumObjectsByKey(sumRecovered, recovered);
          }

          return {
            region: strRegion,
            country: elm["Country/Region"],
            lat: elm["Lat"],
            lng: elm["Long"],
            confirmed: getLastData(cases),
            deaths: getLastData(deaths),
            recovered: getLastData(recovered),
            newConfirmed: getLastDiffData(cases),
            newDeaths: getLastDiffData(deaths),
            newRecovered: getLastDiffData(recovered),
            timeline: {
              cases,
              deaths,
              recovered,
            },
          };
        })
        .filter((elm) => !!elm);
      writeResToJSON(
        {
          data: {
            region: "global",
            todayConfirmed: getLastData(sumCases),
            todayDeath: getLastData(sumDeaths),
            todayRecover: getLastData(sumRecovered),
            newConfirmed: getLastDiffData(sumCases),
            newDeaths: getLastDiffData(sumDeaths),
            newRecovered: getLastDiffData(sumRecovered),
            timeline: {
              cases: sumCases,
              deaths: sumDeaths,
              recovered: sumRecovered,
            },
          },
        },
        "global"
      );
      // TODO: perf 分國家
      // 一個總的統計 for 畫圖
      writePartialTimeSeriesForAPI(finalTimeSeriesData);
      const finalTimeSeriesDataWithoutTimeline = finalTimeSeriesData.map(
        function (element) {
          // 拿掉 timeline 減少大小
          if (element.timeline !== undefined) {
            delete element.timeline;
          }
          return element;
        }
      );
      writeResToJSON(
        {
          data: finalTimeSeriesDataWithoutTimeline,
        },
        "finalTimeSeriesData"
      );
    })
  );

instanceGithub.get("/time_series_covid19_confirmed_US.csv").then((resCases) => {
  const jsonCases = csv2JSON(resCases.data);
  const finalData = jsonCases
    .map((elm) => {
      let cases = removeProperty(elm);
      return {
        region: elm["Combined_Key"],
        country: elm["Country_Region"],
        lat: elm["Lat"],
        lng: elm["Long_"],
        confirmed: getLastData(cases),
      };
    })
    .filter((elm) => elm.confirmed !== "0" && parseInt(elm.confirmed) > 500);
  writeResToJSON({ data: finalData }, "usa");
});
